roachprod: implement fetchlogs

nameisbhaskar opened the pull request:
This implements the fetchlogs feature in roachprod. This was present as a roachtest function and was used to fetch the logs for a roachtest cluster. This is very useful to get all the logs from a cluster. So, this PR moves the same code to roachprod. The destination directory and the timeout for fetching the logs is configurable. The roachtest now invokes the roachprod fetchlogs.

Epic: None
Release note: None

nameisbhaskar committed:
roachprod: implement fetchlogs

This implements the fetchlogs feature in roachprod. This was present as a roachtest function and was used to fetch the logs for a roachtest cluster.
This is very useful to get all the logs from a cluster. So, this PR moves the same code to roachprod. The destination directory and the timeout for fetching the logs is configurable.
The roachtest now invokes the roachprod fetchlogs.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131842)
<!-- Reviewable:end -->

herkolategan reviewed (commented):
Have not noticed a nightly for this one, since it might affect roachtest running a 0.1 prob against it: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/17236584

DarrylWong commented on pkg/cmd/roachprod/main.go:
```diff
@@ -2080,3 +2081,28 @@ Node specification
+			dest = args[1]
+		} else {
+			// trim the node number and keep only the cluster name as prefix of the directory
+			dest = fmt.Sprintf("%s_logs", strings.Split(args[0], ":")[0])
```
nit: can just use `strings.Trim(args[0], ":")` here

DarrylWong replied:
maybe a log message here saying `fmt.Sprintf("Placing logs at %s", dest)`, when there's no destination provided?

DarrylWong reviewed (approved):
Change seems reasonable to me, agree with Herko that a small smoke test doesn't hurt though.

renatolabs commented on pkg/cmd/roachprod/main.go:
```diff
@@ -2080,3 +2081,28 @@ Node specification
+			// trim the node number and keep only the cluster name as prefix of the directory
+			dest = fmt.Sprintf("%s_logs", strings.Split(args[0], ":")[0])
+		}
+		_ = os.Mkdir(dest, 0755)
```
We should probably bail if we can't create the directory.

nameisbhaskar commented:
> Have not noticed a nightly for this one, since it might affect roachtest running a 0.1 prob against it: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/17236584

yes @herkolategan ! I am planning to run a nightly for this before merging.

nameisbhaskar commented:
Thanks @herkolategan @DarrylWong  @renatolabs for reviewing the change. I have addressed the comments. PTAL. I will run the roachtest nightly once to ensure all's good!

nameisbhaskar commented:
> > Have not noticed a nightly for this one, since it might affect roachtest running a 0.1 prob against it: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/17236584
> 
> yes @herkolategan ! I am planning to run a nightly for this before merging.

Seems like you ran it for me. Thanks! I am running it one more time.

nameisbhaskar commented:
bors r=@herkolategan,@renatolabs,@DarrylWong
